jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,343 @@
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+import java.util.function.Predicate;
+
+import org.eclipse.jetty.io.content.ByteBufferContentSource;
+import org.eclipse.jetty.io.content.InputStreamContentSource;
+import org.eclipse.jetty.io.content.PathContentSource;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.resource.MemoryResource;
+import org.eclipse.jetty.util.resource.Resource;
+
+/**
+ * Common IO operations for {@link Resource} content.
+ */
+public class IOResources
+{
+    /**
+     * Gets a {@link Content.Source} with the contents of a resource, if possible.
+     * Non-existent and directory resources have no content, so calling this method on such resource
+     * throws {@link IllegalArgumentException}.
+     *
+     * @param resource the resource from which to get a {@link Content.Source}.
+     * @param bufferPool the {@link ByteBufferPool} to get buffers from. null means allocate new buffers as needed.
+     * @param bufferSize the size of the buffer to be used for the copy. Any value &lt; 1 means use a default value.
+     * @param direct the directness of the buffers, this parameter is ignored if {@code bufferSize} is &lt; 1.
+     * @return the {@link Content.Source}.
+     */
+    public static Content.Source asContentSource(Resource resource, ByteBufferPool bufferPool, int bufferSize, boolean direct)
+    {
+        if (resource.isDirectory() || !resource.exists())
+            throw new IllegalArgumentException("Resource must exist and cannot be a directory: " + resource);
+
+        // Try to find an optimized content source.
+        Path path = resource.getPath();
+        if (path != null)
+        {
+            PathContentSource pathContentSource = new PathContentSource(path, bufferPool);
+            if (bufferSize > 0)
+            {
+                pathContentSource.setBufferSize(bufferSize);
+                pathContentSource.setUseDirectByteBuffers(direct);
+            }
+            return pathContentSource;
+        }
+        if (resource instanceof MemoryResource memoryResource)
+        {
+            byte[] bytes = memoryResource.getBytes();
+            return new ByteBufferContentSource(ByteBuffer.wrap(bytes));
+        }
+
+        // Fallback to wrapping InputStream.
+        try
+        {
+            return new InputStreamContentSource(resource.newInputStream());
```

Comment:
newInputStream can be null, so we should probably check for that and fail nicely

Also need to javadoc how this fails. Can it return null? or will it always throw.

Reply:
I think throwing is the right thing to do here. I've added that check and updated javadoc.